exec.go

```diff
@@ -78,85 +73,49 @@ func ExecCommand(ui Ui, input ExecCommandInput) {
 	if writeEnv {
 		ui.Debug.Println("Writing temporary credentials to ENV")
 
-		env = overwriteEnv(env, "AWS_ACCESS_KEY_ID", val.AccessKeyID)
-		env = overwriteEnv(env, "AWS_SECRET_ACCESS_KEY", val.SecretAccessKey)
+		env.Set("AWS_ACCESS_KEY_ID", val.AccessKeyID)
+		env.Set("AWS_SECRET_ACCESS_KEY", val.SecretAccessKey)
 
 		if val.SessionToken != "" {
-			env = overwriteEnv(env, "AWS_SESSION_TOKEN", val.SessionToken)
-			env = overwriteEnv(env, "AWS_SECURITY_TOKEN", val.SessionToken)
-		}
-	}
-
-	cmd := exec.Command(input.Command, input.Args...)
-	cmd.Env = env
-	cmd.Stdin = os.Stdin
-	cmd.Stdout = os.Stdout
-	cmd.Stderr = os.Stderr
-
-	go func() {
-		sig := <-input.Signals
-		if cmd.Process != nil {
-			cmd.Process.Signal(sig)
-		}
-	}()
-
-	var waitStatus syscall.WaitStatus
-	if err := cmd.Run(); err != nil {
-		if err != nil {
-			ui.Error.Println(err)
-		}
-		if exitError, ok := err.(*exec.ExitError); ok {
-			waitStatus = exitError.Sys().(syscall.WaitStatus)
-			os.Exit(waitStatus.ExitStatus())
+			env.Set("AWS_SESSION_TOKEN", val.SessionToken)
+			env.Set("AWS_SECURITY_TOKEN", val.SessionToken)
 		}
 	}
-}
 
-// write out a config excluding role switching keys
-func writeTempConfig(profile string, conf profiles) (*os.File, error) {
-	tmpConfig, err := ioutil.TempFile(os.TempDir(), "aws-vault")
+	path, err := exec.LookPath(input.Command)
 	if err != nil {
-		return nil, err
+		ui.Error.Fatal(err)
 	}
 
-	newConfig := map[string]string{}
+	pid, err := syscall.ForkExec(path, append([]string{input.Command}, input.Args...), &syscall.ProcAttr{
+		Env:   env,
+		Files: []uintptr{0, 1, 2},
+	})
```

Comment:
```
$ git checkout v3.0.0-dev
HEAD is now at 546d8ef... Add warning about nested exec sessions
$ go build && ./aws-vault exec pda -- pstree $$
-+= 24359 pda -zsh
 \-+= 26285 pda ./aws-vault exec pda -- pstree 24359
   \-+- 26287 pda pstree 24359
     \--- 26288 root ps -axwwo user,pid,ppid,pgid,command
```

Compared to:

```
$ ./aws-vault-2.2.1 exec pda -- pstree $$
-+= 24359 pda -zsh
 \-+= 26391 pda pstree 24359
   \--- 26392 root ps -axwwo user,pid,ppid,pgid,command
```

The process title can be visually tested with:

```sh
./aws-vault-2.2.1 exec pda -- sleep 2  # process title changes to `sleep`
./aws-vault exec pda -- sleep 2        # process title remains `aws-vault`
```

(as reported by tmux window name or iTerm/Terminal tab name)

> My understanding was that ForkExec is like Exec, but forks the current process to the background.

I believe `ForkExec` is how e.g. a shell works; parse the command, fork the current shell process, exec the program in the child process with env updated and new argv. So the executed command remains a child process.

What you're aiming for would be more like this:

* Fork the go process.
* Launch the metadata server in the child process.
* Disown the child server process; reparent to PID 1?
* Exec the user-specified command to replace the original process with a new one.

Thus the originally launched PID becomes the user-specified program after aws-vault has modified the environment and started another process. However that's not possible because Go can't fork itself; it can only fork to exec other commands.

So; in normal mode we could configure the environment then `syscall.Exec()` to replace the process with the user command, and in server mode we could use `syscall.ForkExec` or `exec.Command().Run()` to run the user command as a child process, and miss out on things like process title, simplified signal handling, exit status propagation etc.